applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_truss_element_3D2N.cpp

```diff
@@ -0,0 +1,332 @@
+        const SizeType num_GP = (mpPrimalElement->GetGeometry().IntegrationPoints()).size();
+        if ( (rOutput.size1() != num_dofs) || (rOutput.size2() != num_GP ) )
+            rOutput.resize(num_dofs, num_GP);
+
+        double derivative_pre_factor;
+        this->GetDerivativePreFactor(derivative_pre_factor, rCurrentProcessInfo);
+
+        Vector length_derivative_vector;
+        this->CalculateCurrentLengthDisplacementDerivative(length_derivative_vector);
+
+        for(IndexType i = 0; i < num_dofs; ++i)
+        {
+             for(IndexType j = 0; j < num_GP; ++j)
+                rOutput(i, j) = length_derivative_vector[i] * derivative_pre_factor;
+        }
+    }
+    else
+        KRATOS_ERROR << "Stress displacement derivative only available for Gauss-points quantities!" << std::endl;
+
+    KRATOS_CATCH("")
+}
+
+int AdjointFiniteDifferenceTrussElement::Check(const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY
+
+    KRATOS_ERROR_IF_NOT(mpPrimalElement) << "Primal element pointer is nullptr!" << std::endl;
+
+    //TODO: Check() of primal element should be called, but is not possible because of DOF check!
+
+    KRATOS_ERROR_IF(this->GetGeometry().WorkingSpaceDimension() != 3 || this->GetGeometry().size() != 2)
+        << "The truss element works only in 3D and with 2 noded elements" << "" << std::endl;
+
+    CheckVariables();
+    CheckDofs();
+    CheckProperties(rCurrentProcessInfo);
+
+    KRATOS_ERROR_IF(this->GetGeometry().Length() < std::numeric_limits<double>::epsilon())
+        << "Element #" << this->Id() << " has a length of zero!" << std::endl;
+
+    return 0;
+
+    KRATOS_CATCH("")
+}
+
+
+void AdjointFiniteDifferenceTrussElement::CheckVariables()
+{
+    KRATOS_CHECK_VARIABLE_KEY(DISPLACEMENT);
+    KRATOS_CHECK_VARIABLE_KEY(VELOCITY);
+    KRATOS_CHECK_VARIABLE_KEY(ACCELERATION);
+    KRATOS_CHECK_VARIABLE_KEY(DENSITY);
+    KRATOS_CHECK_VARIABLE_KEY(CROSS_AREA);
+    KRATOS_CHECK_VARIABLE_KEY(ADJOINT_DISPLACEMENT);
+}
+
+void AdjointFiniteDifferenceTrussElement::CheckDofs()
+{
+    const SizeType number_of_nodes = this->GetGeometry().size();
+
+    for (IndexType i = 0; i < number_of_nodes; ++i)
+    {
+        const NodeType &r_node = this->GetGeometry()[i];
+
+        KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(DISPLACEMENT, r_node);
+        KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(ADJOINT_DISPLACEMENT, r_node);
+
+        KRATOS_CHECK_DOF_IN_NODE(ADJOINT_DISPLACEMENT_X, r_node);
+        KRATOS_CHECK_DOF_IN_NODE(ADJOINT_DISPLACEMENT_Y, r_node);
+        KRATOS_CHECK_DOF_IN_NODE(ADJOINT_DISPLACEMENT_Z, r_node);
+    }
+}
+
+void AdjointFiniteDifferenceTrussElement::CheckProperties(const ProcessInfo& rCurrentProcessInfo)
+{
+    const double numerical_limit = std::numeric_limits<double>::epsilon();
+    const PropertiesType & r_properties = GetProperties();
+
+    KRATOS_ERROR_IF(r_properties.Has(CROSS_AREA) == false || r_properties[CROSS_AREA] <= numerical_limit)
```

Comment:
minor: `!r_properties.Has(CROSS_AREA)` instead of `r_properties.Has(CROSS_AREA) == false` would be shorter